cpp/get_dangling_title_references.cc

```diff
@@ -27,106 +27,73 @@
 
 [[noreturn]] void Usage() {
-    std::cerr << "Usage: " << ::progname << " [--patch-to-k10plus] marc_input dangling_log k10plus_concordance\n\n";
-    std::exit(EXIT_FAILURE);
+    ::Usage("[--consider-only-reviews] marc_input dangling_log");
 }
 
-} // unnamed namespace
-
 
 void CollectAllPPNs(MARC::Reader * const reader, std::unordered_set<std::string> * const all_ppns) {
    while (const auto record = reader->read())
       all_ppns->emplace(record.getControlNumber());
 }
 
-bool IsPartOfTitleData(const std::unordered_set<std::string> &all_ppns, const std::string &referenced_ppn) {
+
+inline bool IsPartOfTitleData(const std::unordered_set<std::string> &all_ppns, const std::string &referenced_ppn) {
     return all_ppns.find(referenced_ppn) != all_ppns.cend();
 }
 
 
-void CheckCrossReferences(MARC::Reader * const reader, const std::unordered_set<std::string> &all_ppns,
-                          std::unique_ptr<File> &dangling_log, bool patch_to_k10plus, std::unordered_map<std::string, std::string> &concordance_map) {
-     int unreferenced_ppns(0);
+const std::vector<MARC::Tag> REFERENCE_FIELDS{ MARC::Tag("775"), MARC::Tag("776"), MARC::Tag("780"),
+                                               MARC::Tag("785"), MARC::Tag("787") };
+
+
+void FindDanglingCrossReferences(MARC::Reader * const reader, const bool consider_only_reviews,
+                                 const std::unordered_set<std::string> &all_ppns, File * const dangling_log)
+{
+     unsigned unreferenced_ppns(0);
      while (const auto record = reader->read()) {
-        for (auto &field : record) {
+        for (const auto &field : record) {
             std::string referenced_ppn;
-            if (MARC::IsCrossLinkField(field, &referenced_ppn, REFERENCE_FIELDS)) {
-                if (not IsPartOfTitleData(all_ppns, referenced_ppn)) {
-                    std::string ppn = record.getControlNumber();
-                    if (patch_to_k10plus) {
-                        auto new_ppn = concordance_map.find(ppn);
-                        if (new_ppn == concordance_map.end())
-                            LOG_ERROR("Could not find K10plus-PPN for PPN " + ppn);
-                        auto new_referenced_ppn = concordance_map.find(referenced_ppn);
-                        if (new_referenced_ppn == concordance_map.end()) {
-                            LOG_WARNING("Could not find K10plus-PPN for referenced PPN " + referenced_ppn + " [PPN: " + ppn + "]");
-                            *dangling_log << new_ppn->second << ", K10+ PPN DOES NOT EXIST FOR \"" << referenced_ppn + "\"\n";
-                        }
-                        *dangling_log << new_ppn->second << "," << new_referenced_ppn->second << '\n';
-                    } else
-                        *dangling_log << ppn << "," << referenced_ppn << '\n';
-
-                    ++unreferenced_ppns;
-                }
+            if ((not consider_only_reviews or record.isReviewArticle())
```

Comment:
the first condition could be checked once for the whole record instead of checking it again and again for every single field

Reply:
Done